src/domains/template_generator_base.cpp

```diff
@@ -715,3 +697,40 @@ template_generator_baset::filter_array_domain(const var_specst &var_specs)
   }
   return new_var_specs;
 }
+
+/// Replace all variables in the given index expressions by their loop-back
```

Comment:
❓ What's the rationale behind this?

Reply:
When doing array segmentation, the segment borders may be generic expressions, possibly (even likely) containing variables. We have to determine which SSA versions of the variables to use - if the variable is updated within the same loop as the array, we use its loop-back variant, if it is not, we use its RHS variant (which is the same in the entire loop as it doesn't change).

For instance, when doing `arr[x + i]` in a loop, where `x` is constant within the loop while `i` changes, the segment border would be `x#rhs + i#lb`.